pwndbg/heap_tracking.py

```diff
@@ -0,0 +1,357 @@
+
+            print(f"Chunks in heap {lo_heap.start:#x}-{lo_heap.end:#x}:")
+            for ch in lo_heap:
+                # Check for range overlap.
+                ch_lo_addr = ch.address
+                ch_hi_addr = ch.address + ch.size
+
+                print(f"        {ch.address:#x} ({ch.real_size} bytes)", end="")
+
+                lo_in_range = ch_lo_addr < hi_addr
+                hi_in_range = ch_hi_addr > lo_addr
+
+                if not lo_in_range or not hi_in_range:
+                    # No range overlap.
+                    print(" (out of range)")
+                    continue
+
+                # Check if the chunk is free.
+                found = False
+                for b in bins_list:
+                    if b.contains_chunk(ch.real_size, ch.address):
+                        # The chunk is free. Add it to the free list and install
+                        # a new watch point for it.
+                        nch = Chunk(ch.address, ch.size, ch.real_size, 0)
+                        wp = FreeChunkWatchpoint(nch, self)
+
+                        print(" (free)")
+
+                        self.free_chunks[ch.address] = nch
+                        self.free_wps[ch.address] = wp
+
+                        # Move on to the next chunk.
+                        found = True
+                        break
+                if not found:
+                    print(" (in use)")
+
+        self.alloc_chunks[chunk.address] = chunk
+
+    def free(self, address) -> bool:
+        if address not in self.alloc_chunks:
+            return False
+        chunk = self.alloc_chunks[address]
+        wp = FreeChunkWatchpoint(chunk, self)
+
+        self.free_chunks[chunk.address] = chunk
+        self.free_wps[chunk.address] = wp
+
+        return True
+
+
+class MallocEnterBreakpoint(gdb.Breakpoint):
+    def __init__(self, tracker):
+        super().__init__(MALLOC_NAME, internal=True)
+        self.tracker = tracker
+
+    def stop(self):
+        pwndbg.gdblib.regs.__getattr__.cache.clear()
+        requested_size = pwndbg.arguments.argument(0)
+
+        self.tracker.enter_memory_management()
+        MallocExitBreakpoint(self.tracker, requested_size)
+        return False
+
+
+class MallocExitBreakpoint(gdb.FinishBreakpoint):
+    def __init__(self, tracker, requested_size):
+        super().__init__(internal=True)
+        self.requested_size = requested_size
+        self.tracker = tracker
+
+    def stop(self):
+        pwndbg.gdblib.regs.__getattr__.cache.clear()
+        if not in_program_code_stack():
+            # Untracked.
+            self.tracker.exit_memory_management()
+            return False
+
+        ret_ptr = int(self.return_value)
+        if ret_ptr == 0:
+            # No change.
+            self.tracker.exit_memory_management()
+            return False
+
+        ty = pwndbg.gdblib.typeinfo.ppvoid
+        size = int(pwndbg.gdblib.memory.poi(ty, ret_ptr - ty.sizeof))
+
+        flags = size & 7
```

Comment:
A chance to avoid using magic numbers with the `pwndbg.heap.ptmalloc.Chunk` class, though it lacks a `requested_size` field.

Reply:
I think relying on that class might be a bit much for what I'm trying to do there. Might a `FLAGS_BITMASK = 7` value make what I'm trying to do there clearer?